feat: Issue619 determine the method for generating batches efficiently

mohdt786 opened the pull request:
Closes issue #619, #662
Added Partition support to generate multiple yaml config files

New: 

**- Arguments - file: cli_tools.py**

    1. New command 'generate-partitions' added to generate partitions for the following Validation types:
        1. row 
    2. --partition-type: Specify the type of partition logic:
        1. primary_key
        2. primary_key_mod(TODO)
        3. hash_mod(TODO)
    3. --partition-num: Number of partitions/config files to create.
        Range=[1,1000]
        If specified value is greater than count(*), value if coalesced to count(*)
    4. --config-dir: Directory Path to store YAML Config Files
    5. Added required arguments group to distinguish from optional arguments
    6. Added mutually exclusive arguments group for --hash and --concat
    Example: data-validation generate-partitions row \
                -sc BQ_CONN \
                -tc BQ_CONN \ 
                -tbls bigquery-public-data.new_york_citibike.citibike_stations,mohammedturky-sql.dvt.citibike_stations \   
                --primary-keys station_id,region_id \ 
                --hash \* \
                --filter-status fail \
                --filters 'station_id>3000:station_id>3000' \
                --config-dir partitions_dir \
                --partition-type primary_key \
                --partition-num 20

    For details, use: data-validation generate-partitions -h

**- Partition methods - file: __main__.py**
```
partition_and_store_config_files(): Build ConfigManager and Route based on Validate command
```
**- Partition methods - file: partition_builder.py**

```
1. _get_arg_partition_type(): Get config Directory from args
2. _get_arg_partition_type(): Get partition type from args
3. _get_yaml_from_config(): Convert ConfigManager object to YAML
4. partition_configs(): Route according to partition type
5. _get_primary_key_partition_filters(): Get partition filters for Primary key type partition logic
6. _add_partition_filters_and_store(): Add partition filters to generate multiple yamls and store
```
**- Partition methods - file: partition_row_builder.py**
```
1. _compile_query(): Return Ibis Query object
2. get_max_query(): Build max query
3. get_min_query(): Build min query
4. get_count_query(): Build count query
```

**- Partition methods - file: cli_tools.py**

    get_target_table_folder_path(config_dir, target_folder_name): Create and return target directory

**- Partition methods - file: state_manager.py**

    create_partition_config_directory(config_dir: str,target_folder_name: str): Create target directory for each table

**- Type Hints and Doc string:**

    1. Added Type Hints to the above methods
    7. Added Doc string with desc, args and return type for above methods

mohdt786 committed:
init

mohdt786 committed:
init

mohdt786 committed:
init

mohdt786 committed:
init

mohdt786 committed:
Added Function: build_and_store_config_files

mohdt786 committed:
feat: Added Partition support to generate multiple yaml config files

New:
    Arguments - file: cli_tools.py
    1. New command 'get-partitions' added to generate partitions for the following Validation types:
        1. row
        2. custom-query(TODO)
    2. --partition-type: Specify the type of partition logic:
        1. primary_key
        2. primary_key_mod(TODO)
        3. hash_mod(TODO)
    3. --partition-num: Number of partitions/config files to create.
        Range=[1,1000]
        If specified value is greater than count(*), value if coalesced to count(*)
    4. --config-dir: Directory Path to store YAML Config Files
    5. Added required arguments group to distinguish from optional arguments
    6. Added mutually exclusive arguments group for --hash and --concat

    Constants - file: consts.py
    1. Added DEFAULT_PARTITION_TYPE
    2. Added PARTITION_TYPES

    Partition methods - file: __main__.py
    1. _get_arg_partition_type(args): extract and return partition logic
    2. partition_and_store_config_files(args): Build/split config managers and store yaml files
    3. partition_configs(args, config_managers): Create a list of lists of config managers using partition filters
    4. _get_primary_key_partition_filters(args, config_manager): Get filters for primary_key partition logic
    5. _add_partition_filters_to_config(config_managers, partition_filters): Split ConfigManager objects and Add partition Filters
    6. get_dataframe(config_manager): Build source and target pandas dataframes from input ConfigManager object
    7. build_primary_key_agg_config_managers_from_args(args): Build a list of ConfigManager object for finding count, min and max of primary_key

    Partition methods - file: data_validation.py
    1. get_pandas_df(): Build source and target queries, return source and target dataframes

    Type Hints and Doc string:
    1. Added Type Hints to the above methods
    2. Added Doc string with desc, args and return type for above methods

mohdt786 committed:
Merge branch 'develop' into issue619-determine-the-method-for-generating-batches-efficiently

mohdt786 committed:
fix: Added Partition support to generate multiple yaml config files

New:

    Partition methods - file: __main__.py
    1. _add_partition_filters_and_store(config_managers, partition_filters,config_dir,args): Split ConfigManager objects, Add partition Filters and store in target dir
    2. _get_arg_config_dir(args): Return String yaml config folder pathfrom args.

    Partition methods - file: cli_tools.py
    1. get_target_table_folder_path(config_dir, target_folder_name): Create and return target directory

    Partition methods - file: state_manager.py
    1. create_partition_config_directory(config_dir: str,target_folder_name: str)

    Type Hints and Doc string:
    1. Added Type Hints to the above methods
    2. Added Doc string with desc, args and return type for above methods

mohdt786 committed:
testing and linting

mohdt786 committed:
fix: linting and test

mohdt786 committed:
fix: Added Partition support to generate multiple yaml config files

New:

    Partition methods - file: __main__.py
    1. _add_partition_filters_and_store(config_managers, partition_filters,config_dir,args): Split ConfigManager objects, Add partition Filters and store in target dir
    2. _get_arg_config_dir(args): Return String yaml config folder pathfrom args.

    Partition methods - file: cli_tools.py
    1. get_target_table_folder_path(config_dir, target_folder_name): Create and return target directory

    Partition methods - file: state_manager.py
    1. create_partition_config_directory(config_dir: str,target_folder_name: str)

    Type Hints and Doc string:
    1. Added Type Hints to the above methods
    2. Added Doc string with desc, args and return type for above methods

mohdt786 committed:
feat: Added Partition support to generate multiple yaml config files

New:      Arguments - file: cli_tools.py     1. New command 'get-partitions' added to generate partitions for the following Validation types:         1. row          2. custom-query(TODO)     2. --partition-type: Specify the type of partition logic:         1. primary_key         2. primary_key_mod(TODO)         3. hash_mod(TODO)     3. --partition-num: Number of partitions/config files to create.         Range=[1,1000]         If specified value is greater than count(*), value if coalesced to count(*)     4. --config-dir: Directory Path to store YAML Config Files     5. Added required arguments group to distinguish from optional arguments     6. Added mutually exclusive arguments group for --hash and --concat     Example: data-validation get-partitions row \                 -sc BQ_CONN \                 -tc BQ_CONN \                  -tbls bigquery-public-data.new_york_citibike.citibike_stations,mohammedturky-sql.dvt.citibike_stations \                   "--primary-keys station_id,region_id \                  --hash * \                 --filter-status fail \                 --filters 'station_id>3000:station_id>3000' \                 --config-dir partitions_dir \                 --partition-type primary_key \                 --partition-num 20      Constants - file: consts.py     1. Added DEFAULT_PARTITION_TYPE     2. Added PARTITION_TYPES      Partition methods - file: __main__.py     1. _get_arg_partition_type(args): extract and return partition logic     2. partition_and_store_config_files(args): Build/split config managers and store yaml files     3. partition_configs(args, config_managers): Create a list of lists of config managers using partition filters     4. _get_primary_key_partition_filters(args, config_manager): Get filters for primary_key partition logic     5. _add_partition_filters_and_store(config_managers, partition_filters,config_dir,args): Split ConfigManager objects, Add partition Filters and store in target dir     6. get_dataframe(config_manager): Build source and target pandas dataframes from input ConfigManager object     7. build_primary_key_agg_config_managers_from_args(args): Build a list of ConfigManager object for finding count, min and max of primary_key     8. _get_arg_config_dir(args): Return String yaml config folder pathfrom args.       Partition methods - file: data_validation.py     1. get_pandas_df(): Build source and target queries, return source and target dataframes      Partition methods - file: cli_tools.py     1. get_target_table_folder_path(config_dir, target_folder_name): Create and return target directory      Partition methods - file: state_manager.py     1. create_partition_config_directory(config_dir: str,target_folder_name: str)      Type Hints and Doc string:     1. Added Type Hints to the above methods     2. Added Doc string with desc, args and return type for above methods

New Command added - 'get-partitions'

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: removed type hint for method - get_pandas_df

mohdt786 commented:
/gcbrun

mohdt786 committed:
linting and fixes

mohdt786 commented:
/gcbrun

mohdt786 committed:
type hints and lint

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: Raise value error for:

Validation type: custom-query
Partition type: primary_key_mod & hash_mod

mohdt786 commented:
/gcbrun

mohdt786 commented:
cc: @Raniksingh

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -331,6 +363,176 @@ def _configure_validate_parser(subparsers):
+        help="Comma separated list of columns for concat 'col_a,col_b' or * for all columns",
+    )
+
+    row_partition_parser.add_argument(
```
Comparison fields should also be in the mutually_exclusive_arguments since a user can only do hash, concat, or comp-fields

mohdt786 replied:
Done!

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -169,9 +170,40 @@ def configure_arg_parser():
+def _configure_get_partition_parser(subparsers):
+    """Configure arguments to generate partitioned config files."""
+    validate_parser = subparsers.add_parser(
+        "get-partitions",
```
Can we rename this 'generate-partitions'?

mohdt786 replied:
Yes sounds better.
Done!

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -169,9 +170,40 @@ def configure_arg_parser():
 
+def _configure_get_partition_parser(subparsers):
+    """Configure arguments to generate partitioned config files."""
+    validate_parser = subparsers.add_parser(
```
Can we rename this partition_parser? This is confusing since we have a validate_parser for the `data-validation validate` command

mohdt786 replied:
Yes.
Done!

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -169,9 +170,40 @@ def configure_arg_parser():
+        "--result-handler-config", "-rc", help="Result handler config details"
+    )
+
+    validate_subparsers = validate_parser.add_subparsers(
```
See above - let's rename this to partition_subparsers

mohdt786 replied:
Done!

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -331,6 +363,176 @@ def _configure_validate_parser(subparsers):
+        type=threshold_float,
+        help="Float max threshold for percent difference",
+    )
+    row_partition_parser.add_argument(
```
We can't have grouped-columns with row validation

mohdt786 replied:
`grouped-columns` was part of the previous parser, hence had added the same arguments
```    
row_parser.add_argument(
        "--grouped-columns",
        "-gc",
        help="Comma separated list of columns to use in GroupBy 'col_a,col_b'",
    )
```

mohdt786 replied:
Removed `grouped-columns` from arguments for row validation in `row_partition_parser`

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -331,6 +363,176 @@ def _configure_validate_parser(subparsers):
+    _add_common_partition_arguments(row_partition_parser, required_arguments)
+
+
+def _configure_custom_query_partition_parser(custom_query_partition_parser) -> None:
```
I think we shouldn't add a custom_query partition parser in this PR if we don't plan on supporting it yet

mohdt786 replied:
Yes, makes sense
Removed custom_query partition parser

nehanene15 commented on data_validation/state_manager.py:
```diff
@@ -220,3 +221,26 @@ def _list_gcs_directory(self, directory_path: str) -> List[str]:
 
         return blobs
+
+    def create_partition_config_directory(
```
If the user supplies a directory with 'gs://' we can create the directory in GCS, otherwise it will be local. PR #654 adds support for reading from a GCS or local directory that you can see as an example

mohdt786 replied:
While creating connections, user specifies `PSO_DV_CONFIG_HOME`, if user specifies it as a GCS path, connection details are saved on GCS. Moreover, the YAML Config files are then also saved in the same path.

Reference: https://github.com/GoogleCloudPlatform/professional-services-data-validator/blob/develop/docs/connections.md

> The connections can also be stored in GCS using PSO_DV_CONFIG_HOME. To do so simply add the GCS path to the environment. Note that if this path is set, query validation configs will also be saved here.
> 
> eg. export PSO_DV_CONFIG_HOME=gs://my-bucket/my/connections/path/

Hence, to save the partitions, user only needs to specify the Directory name, and the location will be decided by `PSO_DV_CONFIG_HOME`

mohdt786 replied:
Do we need a separate check, which, irrespective of `PSO_DV_CONFIG_HOME` decides the location of Directory based on the path supplied by the user at `--config-dir`?

nehanene15 commented on data_validation/__main__.py:
```diff
@@ -472,10 +501,284 @@ def store_yaml_config_file(args, config_managers):
+        A list of type ConfigManager to get count, min and max of primary_key column.
+    """
+    agg_config_managers = []
+    config_type = consts.COLUMN_VALIDATION
```
Why is this being set to column validation?

nehanene15 commented on data_validation/__main__.py:
```diff
@@ -472,10 +501,284 @@ def store_yaml_config_file(args, config_managers):
+    source_client = clients.get_data_client(mgr.get_connection_config(args.source_conn))
+    target_client = clients.get_data_client(mgr.get_connection_config(args.target_conn))
+
+    format = "table"
```
We should be getting all these variables from the args parameters

nehanene15 commented on data_validation/__main__.py:
```diff
@@ -472,10 +501,284 @@ def store_yaml_config_file(args, config_managers):
+
+        primary_key = primary_keys[0]
+        aggregate_configs = [config_manager.build_config_count_aggregate()]
+        aggregate_configs += config_manager.build_config_column_aggregates(
```
Why are we building column aggregates?

nehanene15 commented on data_validation/__main__.py:
```diff
@@ -472,10 +501,284 @@ def store_yaml_config_file(args, config_managers):
+    )
+
+
+def _get_primary_key_partition_filters(args: Namespace) -> List[List[str]]:
```
We want to build this query using Ibis, not pandas ideally. This logic should also go in a new file query_builder/partitioned_row_builder.py or something similar. It will be very similar to random_row_builder where it [builds the query](https://github.com/GoogleCloudPlatform/professional-services-data-validator/blob/develop/data_validation/query_builder/random_row_builder.py#L110) using Ibis.

For example, in Ibis you can do table.count() or table[primary_key].min() and then execute the query on source or target.

nehanene15 reviewed (commented):
The biggest suggestion is to use Ibis instead of pandas to generate the partitions similar to the pattern in random row builder

mohdt786 committed:
Merge branch 'develop' into issue619-determine-the-method-for-generating-batches-efficiently

mohdt786 committed:
fix: Updated arguments and argument parser

1. Changed command from get_partitions to generate partitions
2. Renamed validate_parser to partition_parser
3. Renamed validate_subparser to partition_subparser
4. Removed --grouped-columns argument for row validation
5. Removed custom_query partition parser since it is not supported in this PR

mohdt786 committed:
feat: Added Partition support to generate multiple yaml config files

Reconfigured and shifted Partition methods to data_validation/query_builder/partition_row_builder.py and data_validation/partition_builder.py

file: partition_row_builder.py
1. Build PartitionRowBuilder object
2. Get Ibis queries for Min, Max and Count for primary key column

file: partition_builder.py
1. Build PartitionBuilder object
2. Create ConfigManager objects for input tables
3. Generate Partition filters
4. Save partitions in input directory as multiple YAMLs

mohdt786 committed:
Linting and TypeHint added

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: Added Warning logs and Documentation

1. Warning logs in case of Source/Target count, min or max mismatch. 2. Raise TypeError when primary key is not numeric

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: removed unwanted Docs

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: Added --tables-list to required for generate-partitions row command

mohdt786 committed:
Merge branch 'develop' into issue619-determine-the-method-for-generating-batches-efficiently

mohdt786 commented:
/gcbrun

nehanene15 commented on .gitignore:
```diff
@@ -84,7 +84,14 @@ terraform.rc
 # Test temp files
 source_table_data.json
 target_table_data.json
+data_validation/cli_tools.py
```
We should remove this since we usually want to include changes to cli_tools

mohdt786 replied:
Done

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -168,9 +172,34 @@ def configure_arg_parser():
+        help="Type of Validation.",
+    )
+
+    row_partition_parser = partition_subparsers.add_parser(
```
I think we should only support row validations. Column validations aggregate results so memory isn't an issue (i.e count * or sum col_a will always return one row as a result, but row validations return multiple).

We can get rid of this subparser: `data-validation generate-table-partitions -sc bq -tc bq ...`

mohdt786 replied:
Done

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -168,9 +172,34 @@ def configure_arg_parser():
+def _configure_partition_parser(subparsers):
+    """Configure arguments to generate partitioned config files."""
+    partition_parser = subparsers.add_parser(
+        "generate-partitions",
```
Maybe we can change this to `--generate-table-partitions` so it's easy to understand what it's partitioning and how it's different from random row

mohdt786 replied:
Yes. Done

nehanene15 commented on docs/examples.md:
`````diff
@@ -30,6 +30,12 @@ Above command creates a YAML file named citibike.yaml that can be used to run va
+````shell script
+data-validation generate-partitions row -sc my_bq_conn -tc my_bq_conn -tbls bigquery-public-data.new_york_citibike.citibike_stations --primary-keys station_id --hash '*' --filters 'station_id>3000' -cdir partitions_dir --partition-type primary_key --partition-num 200
+````
+Above command creates multiple partitions based on primary key. Number of configuration files is decided by `--partition-num`
`````
Let's also add the CLI and a description of this feature in the README similar to [this](https://github.com/GoogleCloudPlatform/professional-services-data-validator#row-validations). We can have a new section titled 'Generate Table Partitions for Large Table Row Validations' with the command description.
We should also include that we only support monotonically increasing keys currently

mohdt786 replied:
Done

nehanene15 commented on data_validation/partition_builder.py:
```diff
@@ -0,0 +1,263 @@
+                validation_builder.target_builder,
+            )
+
+            # accepted_datatypes = ("int64", "int32")
```
Should this be uncommented?

mohdt786 replied:
There is a datatype check below, will remove commented block
```
# If Primary key is non numeric, raise Type Error
if not (source_min.is_integer() and target_min.is_integer()):
    raise TypeError(
        f"Supplied Primary key is not of type Numeric: {primary_key}"
    )
```

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -641,6 +733,71 @@ def _add_common_arguments(parser):
+        required=True,
+        help="Directory Path to store YAML Config Files",
+    )
+    required_arguments.add_argument(
```
I think we should get rid of this flag and instead just support a --partition-key flag. This way the partition key and PK don't necessarily need to be the same column (if no --partition-key is supplied, we can use the PK). In the future when we support hashmod or partitionkeymod we can add flags depending on what user input we expect.

mohdt786 replied:
Done. Removed `--partition-type` flag. Default partitions would be generated on `--partition-key`. 
`--partition-key` has been made optional, hence in its absence, it will default to first `--primary-keys`

mohdt786 committed:
Merge branch 'develop' into issue619-determine-the-method-for-generating-batches-efficiently

mohdt786 committed:
refactor: Removed ROW command and added desc to README.md

1. Removed `row` command, default partition logic is row 2. Added `--primary-key` 3. Added to Examples.md 4. Added to readme.md

mohdt786 commented:
/gcbrun

nehanene15 commented on README.md:
````diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
 ```
+#### Generate Table Partitions for Large Table Row Validations
 
+Below is the command syntax for Generating Tabe Partitions for performing Row Validation on large Tables.
````
```suggestion
Below is the command syntax for generating table partitions in order to perform row validations on large tables with memory constraints. 
```

mohdt786 replied:
Done

nehanene15 commented on README.md:
```diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
+#### Generate Table Partitions for Large Table Row Validations
 
+Below is the command syntax for Generating Tabe Partitions for performing Row Validation on large Tables.
+The following command splits Row Validation into multiple steps by generating multiple Config files which can then be run in Parallel or independently.
```
```suggestion
The command generates and stores multiple YAML configs that represent chunks of the large table using filters (`WHERE partition_key > X AND partition_key < Y`). You can then run the configs in the directory serially with the `data-validation configs run --config-dir PATH` command as described [here](https://github.com/GoogleCloudPlatform/professional-services-data-validator#yaml-configuration-files).
```

nehanene15 replied:
Does this PR support GCS based config-dir paths?

mohdt786 replied:
Done

nehanene15 commented on README.md:
```diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
+Below is the command syntax for Generating Tabe Partitions for performing Row Validation on large Tables.
+The following command splits Row Validation into multiple steps by generating multiple Config files which can then be run in Parallel or independently.
+
+The command takes the same parameters as required for `Row Validation` *plus* few commands to implement the partitioning logic.
```
Since we already describe the flags needed below, I don't think we need lines 207-213. We should keep the note that only monotonically increasing keys are supported though.

mohdt786 replied:
Done

nehanene15 commented on README.md:
```diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
+  --hash COLUMNS        Comma separated list of columns to hash or * for all columns 
+  --concat COLUMNS      Comma separated list of columns to concatenate or * for all columns (use if a common hash function is not available between databases)
+  --config-dir CONFIG_DIR, -cdir CONFIG_DIR
+                        Directory Path to store YAML Config Files
```
We should specify if this path is local, GCS, or either.

nehanene15 commented on README.md:
```diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
+                        In case this value exceeds the row count of the source/target table, its will be decreased to max(source_row_count, target_row_count)
+  [--partition-key PARTITION_KEY, -partkey PARTITION_KEY]
+                        Column on which the partitions would be generated. Column type must be integer. Defaults to Primary key
+  [--threshold THRESHOLD, -th THRESHOLD]
```
We shouldn't support this threshold flag as we aren't running a validation and can't compare percent difference.

mohdt786 replied:
Done

nehanene15 commented on README.md:
```diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
+                        Finds a set of random rows of the first primary key supplied.
+  [--random-row-batch-size RANDOM_ROW_BATCH_SIZE, -rbs RANDOM_ROW_BATCH_SIZE]
+                        Row batch size used for random row filters (default 10,000).
+  [--result-handler-config RESULT_HANDLER_CONFIG, -rc RESULT_HANDLER_CONFIG]
```
result-handler-config is deprecated so we shouldn't document it. We just keep it in cli_tools for backwards compatibility

mohdt786 replied:
Done

nehanene15 commented on README.md:
```diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
+                        Row batch size used for random row filters (default 10,000).
+  [--result-handler-config RESULT_HANDLER_CONFIG, -rc RESULT_HANDLER_CONFIG]
+                        Result handler config details
+  [--bq-result-handler or -bqrh PROJECT_ID.DATASET.TABLE]
```
We shouldn't support result handler since we don't write any validation results to BQ.

mohdt786 replied:
Done

nehanene15 commented on README.md:
```diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
+                        Path to SA key file for result handler output
+  [--format FORMAT, -fmt FORMAT]
+                        Set the format for printing command output, Supported formats are (text, csv, json, table). Defaults to table
+  [--filter-status FILTER_STATUS, -fs FILTER_STATUS]
```
We don't need any of the above flags (line 252-265) as we aren't running validations, just creating configs.

nehanene15 commented on README.md:
```diff
@@ -197,7 +197,73 @@ data-validation (--verbose or -v) (--log-level or -ll) validate row
+                        Colon spearated string values of source and target filters.
+                        If target filter is not provided, the source filter will run on source and target tables.
+                        See: *Filters* section
+  [--use-random-row, -rr]
```
Supporting random row here wouldn't make sense since we want to break a full table into chunks. Random row would select rows and add a WHERE clause to select random rows; we want to cater this command to only full table row validations

mohdt786 replied:
Done

nehanene15 commented on data_validation/__main__.py:
```diff
@@ -217,11 +210,15 @@ def build_config_from_args(args, config_manager):
     configs = []
 
-    validate_cmd = args.validate_cmd.capitalize()
+    # Since `generate-table-partitions` is default to `validate_cmd=row`,
```
```suggestion
    # Since `generate-table-partitions` defaults to `validate_cmd=row`,
```

mohdt786 replied:
Done

nehanene15 commented on data_validation/__main__.py:
```diff
@@ -491,10 +488,34 @@ def store_yaml_config_file(args, config_managers):
+    """
+    # Default Validate Type
+    validate_cmd = consts.ROW_VALIDATION
+    config_managers = build_config_managers_from_args(args, validate_cmd)
```
```suggestion
    config_managers = build_config_managers_from_args(args, consts.ROW_VALIDATION)
```

mohdt786 replied:
Done

nehanene15 commented on data_validation/__main__.py:
```diff
@@ -491,10 +488,34 @@ def store_yaml_config_file(args, config_managers):
+        None
+    """
+    # Default Validate Type
+    validate_cmd = consts.ROW_VALIDATION
```
Instead of declaring this separately, we can pass in the constant directly to the next line.

mohdt786 replied:
Done

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -168,9 +172,104 @@ def configure_arg_parser():
+
+    # Group all optional arguments together
+    optional_arguments = partition_parser.add_argument_group("optional arguments")
+    optional_arguments.add_argument(
```
We don't need threshold, use random row, or random row batch size since we aren't running a typical validation

mohdt786 replied:
Done

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -641,6 +740,75 @@ def _add_common_arguments(parser):
+    )
+
+    # Optional arguments
+    optional_arguments.add_argument(
```
Don't need flags from line 770 - 801 as mentioned above

mohdt786 replied:
Removed from readme.md but kept required ones in cli_tools.py to build ConfigManager object

nehanene15 commented on data_validation/consts.py:
```diff
@@ -90,6 +90,8 @@
 ]
 
+# Partition types
+PARTITION_TYPES = ["primary_key", "primary_key_mod", "hash_mod"]
```
We can remove this and line 71 since it's unused for now.

mohdt786 replied:
Done

nehanene15 commented on data_validation/partition_builder.py:
```diff
@@ -0,0 +1,241 @@
+                yaml_configs = self._get_yaml_from_config(config_manager)
+                target_file_name = "0" * (4 - len(str(pos))) + str(pos) + ".yaml"
+                config_file_path = os.path.join(target_folder_path, target_file_name)
+                cli_tools.store_validation(config_file_path, yaml_configs)
```
Each store_validation call logs a success message for the file being written [here](https://github.com/GoogleCloudPlatform/professional-services-data-validator/blob/develop/data_validation/state_manager.py#L175) - if a user has 1000 partitions this would log a lot of times... did you encounter this while testing?

mohdt786 committed:
feat: Added support to save partition configs to GCS bucket

1. Configs can be saved to GCS specifying a full gs:// path
2. Updated README.MD
3. Removed Random Row and Threshold arguments for generate-table-partitions
4. Removed unused consts
5. Modified partition_builder.py
6. Updated state_manager.py to support GCS path independant of PSO_DV_CONFIG_HOME

mohdt786 committed:
Merge branch 'develop' into issue619-determine-the-method-for-generating-batches-efficiently

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: fixed path conflicts

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: re-wrote logs which are used in tests/unit/test_cli_tools.py::test_create_and_list_connections & tests/unit/test_cli_tools.py::test_create_and_list_and_get_validations

mohdt786 commented:
/gcbrun

nehanene15 commented on data_validation/cli_tools.py:
```diff
@@ -646,6 +728,79 @@ def _add_common_arguments(parser):
+        required=True,
+        help="Directory Path to store YAML Config Files. "
+        "GCS: Provide a full gs:// path of the target directory. "
+        "Eg: `gs://<BUCKET>/partiitons_dir`. "
```
```suggestion
        "Eg: `gs://<BUCKET>/partitons_dir`. "
```

mohdt786 replied:
Fixed the typo

nehanene15 commented on data_validation/__main__.py:
```diff
@@ -258,12 +261,18 @@ def build_config_managers_from_args(args):
-    )
+    # Random row validation is not supported for schema validation and
+    # generate-table-partitions
+    use_random_rows = None
```
Instead of doing this, we should just leave the random_rows/random_row_batch_size/threshold as an argument in the partition parser. We should just remove it from documentation so people don't use it.

mohdt786 replied:
Okay, done

mohdt786 committed:
test: Added test cases for partition_builder.py

1. Covered testing PartitionBuilder object 2. Generate Partitions 3. Generate YAML from Configs with new partition filters

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: extra line adjustment in consts.py

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: updated source/target partition_key datatype check to support python3.7

mohdt786 commented:
/gcbrun

mohdt786 committed:
Added test cases for storing partitions in local dir

mohdt786 committed:
Merge branch 'develop' into issue619-determine-the-method-for-generating-batches-efficiently

mohdt786 commented:
/gcbrun

mohdt786 committed:
lint: tests/unit/test_partition_builder.py

mohdt786 commented:
/gcbrun

nehanene15 commented on tests/unit/test_partition_builder.py:
```diff
@@ -0,0 +1,491 @@
+    return data_validation.partition_builder
+
+
+def teardown_module(module):
```
Where do we use this function?

mohdt786 replied:
It is a clean up method called automatically by PyTest at the end of all the test methods. Used here to delete the directory created while testing `test_store_yaml_partitions_local()`

nehanene15 commented on tests/unit/test_partition_builder.py:
```diff
@@ -0,0 +1,491 @@
+    parser = cli_tools.configure_arg_parser()
+    mock_args = parser.parse_args(CLI_ARGS_JSON_SOURCE)
+
+    mock_partition_filters = _generate_fake_partition_filters(
```
Instead of generating fake partition filters, we can just manually check that the partition_filters_list has the correct length/attributes that we expect so we don't have to copy the logic over. 
```
expected_partition_filters_list = ["key >= 0 and key < 10", "key >= 11 and key < 20"]
assert len(partition_filters_list[0]) == 20
assert partition_filters_list[0] == expected_partition_filters_list
```
This can also be done for _generate_fake_yaml_configs() to avoid having any logic in the tests.

mohdt786 replied:
Removed generation of `partition_filters_list` and `yaml_configs_list` via function.

Added expected `PARTITION_FILTERS_LIST` and `YAML_CONFIGS_LIST` to a json file `tests/unit/test_inputs/test_partition_builder.json` since expected `YAML_CONFIGS_LIST` is too large to store in `tests/unit/test_partition_builder.py` and would make it less readable.

nehanene15 reviewed (commented):
LGTM besides a 2 comments on the tests - thanks!

mohdt786 committed:
Added test inputs to tests/unit/test_inputs/test_partition_builder.json

mohdt786 committed:
lint: tests/unit.test_partition_builder.py

mohdt786 commented:
/gcbrun

nehanene15 commented on tests/unit/test_inputs/test_partition_builder.json:
```diff
@@ -0,0 +1,1956 @@
+{
+    "PARTITION_FILTERS_LIST": [
```
This file is really long... can we adjust to 3-5 partitions instead of 20?

mohdt786 replied:
Done. Reduced test input size

mohdt786 committed:
fix: Reduced test input size

mohdt786 commented:
/gcbrun

mohdt786 commented:
/gcbrun

mohdt786 commented:
/gcbrun

mohdt786 committed:
fix: Added logging before writing partitions

mohdt786 commented:
/gcbrun
